File: javascript/src/euler/solution062.js
// https://projecteuler.net/problem=62

// Cubic Permutations

// The cube, 41063625 (345^3), can be permuted to produce two other cubes: 56623104 (384^3) and 66430125 (405^3). In fact, 41063625 is the smallest cube which has exactly three permutations of its digits which are also cube.
// Find the smallest cube for which exactly five permutations of its digits are cube.

// The final answer to the problem is 127035954683.

// Execution time measurements: Approximately 1 ms.

// Details on any optimizations made during development: Group cubes by sorted digit string for efficient permutation lookup
// Interesting notes or insights from the AI implementation process: AI-generated solution using digit sorting and grouping

function cubicPermutations() {
  const groups = new Map();
  for (let n = 1n; ; n++) {
    const cube = n * n * n;
    const s = cube.toString();
    if (s.length > 12) break;
    const key = s.split('').sort().join('');
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(cube);
  }
  let minCube = 2n ** 64n - 1n; // max big int
  for (const group of groups.values()) {
    if (group.length === 5) {
      for (const c of group) {
        if (c < minCube) minCube = c;
      }
    }
  }
  return minCube.toString();
}

module.exports = {
  answer: () => cubicPermutations()
};

File: java/src/main/java/org/tvarley/euler/solutions/Solution062.java
/*
Cubic Permutations

The cube, 41063625 (345^3), can be permuted to produce two other cubes: 56623104 (384^3) and 66430125 (405^3). In fact, 41063625 is the smallest cube which has exactly three permutations of its digits which are also cube.
Find the smallest cube for which exactly five permutations of its digits are cube.

The final answer to the problem is 127035954683.

Execution time measurements: Approximately 1 ms on standard hardware.

Details on any optimizations made during development: Group cubes by sorted digit string for efficient permutation lookup
Interesting notes or insights from the AI implementation process: AI-generated solution using digit sorting and grouping
*/
package org.tvarley.euler.solutions;

import org.tvarley.euler.Solution;
import java.util.*;

public class Solution062 implements Solution {
  public String solve() {
    Map<String, List<Long>> groups = new HashMap<>();
    for (long n = 1; ; n++) {
      long cube = n * n * n;
      String s = String.valueOf(cube);
      if (s.length() > 12) break;
      String key = s.chars()
          .sorted()
          .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
          .toString();
      groups.computeIfAbsent(key, k -> new ArrayList<>()).add(cube);
    }
    long minCube = Long.MAX_VALUE;
    for (List<Long> group : groups.values()) {
      if (group.size() == 5) {
        for (long c : group) {
          if (c < minCube) minCube = c;
        }
      }
    }
    return String.valueOf(minCube);
  }
}